### PR TITLE
Stream documents to be indexed

### PR DESCRIPTION
**Pull request by labkey-matthewb**
#### Rationale
We currently buffer the tika converted documents in memory (the text version of whatever file is being indexed).  These files can be very large.  This change allows for these files to overflow to disk to limit the indexers demands on the java heap.

#### Related Pull Requests
* 

#### Changes
*

### CODE REVIEW

File: api/src/org/labkey/api/util/FileUtil.java [hunk 1]
```diff
@@ -1329,6 +1342,168 @@ else if (!FileUtil.hasCloudScheme(path))
...
+            this.characterLimitInMemory = charBuffer.capacity();
+        }
+
+
+        public Writer getWriter()
+        {
+            // should only ask for _writer once
+            assert null == _writer;
+            // CONSIDER ByteBuffer.allocateDirect(), for now caller can pass in a direct buffer if desired
+            if (null == _charBuffer)
```
Comment (labkey-adam) [inline api/src/org/labkey/api/util/FileUtil.java]: Why not do this allocation in `TempTextFileWrapper(int)`, passing the `charBuffer` to `TempTextFileWrapper(CharBuffer)`?

File: api/src/org/labkey/api/util/FileUtil.java [hunk 2]
```diff
@@ -1329,6 +1342,168 @@ else if (!FileUtil.hasCloudScheme(path))
...
+     * direct to cache the first portion of the file to avoid hitting the
+     * file system if the file is smaller.
+     *
+     * The caller needs to call close() on this object or the Reader returned
+     * by getReader().  Calling close on both is OK.
+     */
+    public static class TempTextFileWrapper implements Closeable
+    {
+        final int characterLimitInMemory;
+        ByteBuffer _byteBuffer = null;
```
Comment (labkey-adam) [inline api/src/org/labkey/api/util/FileUtil.java]: `@Nullable`

File: api/src/org/labkey/api/util/FileUtil.java [hunk 3]
```diff
@@ -1329,6 +1342,168 @@ else if (!FileUtil.hasCloudScheme(path))
...
+        CharBuffer _charBuffer = null;
+//        int _maxPosition = 0;
```
Comment (labkey-adam) [inline api/src/org/labkey/api/util/FileUtil.java]: Remove?

File: api/src/org/labkey/api/util/FileUtil.java [hunk 4]
```diff
@@ -1329,6 +1342,168 @@ else if (!FileUtil.hasCloudScheme(path))
...
+                @Override
+                public void write(@NotNull char[] cbuf, int off, int len) throws IOException
+                {
+                    if (closed)
+                        throw new IOException("Writer is closed");
+                    if (_charBuffer.remaining() > 0)
+                    {
+                        var l = Math.min(_charBuffer.remaining(), len);
+                        _charBuffer.put(cbuf, off, l);
+//                        _maxPosition = Math.max(_maxPosition, _charBuffer.position());
```
Comment (labkey-adam) [inline api/src/org/labkey/api/util/FileUtil.java]: remove?

File: api/src/org/labkey/api/util/FileUtil.java [hunk 5]
```diff
@@ -1329,6 +1342,168 @@ else if (!FileUtil.hasCloudScheme(path))
...
+     * file system if the file is smaller.
+     *
+     * The caller needs to call close() on this object or the Reader returned
+     * by getReader().  Calling close on both is OK.
+     */
+    public static class TempTextFileWrapper implements Closeable
+    {
+        final int characterLimitInMemory;
+        ByteBuffer _byteBuffer = null;
+        CharBuffer _charBuffer = null;
```
Comment (labkey-adam) [inline api/src/org/labkey/api/util/FileUtil.java]: Could be `final` if initialized in the first constructor

**Review by labkey-adam — CHANGES_REQUESTED** [review]
Couple suggestions for `TempTextFileWrapper`

File: api/src/org/labkey/api/util/FileUtil.java (hunk 1 above)
Comment (labkey-matthewb) [reply api/src/org/labkey/api/util/FileUtil.java]: Moving the _charBuffer allocation is makes sense, but I want to make sure the caller only asks for the Writer/Reader once.  I don't want the interface to look like might be more than one writer (or reader) that separately track position.  Using null as a marker is simple.

**Review by labkey-adam — APPROVED** [review]

**Pull request merged**